types/validator_set.go

```diff
@@ -370,52 +371,87 @@ func processChanges(origChanges []*Validator) (updates, removals []*Validator, e
 	return updates, removals, err
 }
 
+type powerChange struct {
+	address Address
+	delta   int64
+}
+
 // Verifies a list of updates against a validator set, making sure the allowed
 // total voting power would not be exceeded if these updates would be applied to the set.
 //
+// Inputs:
+// updates - a list of proper validator changes, i.e. they have been verified by processChanges for duplicates
+//   and invalid values.
+// vals - the original validator set. Note that vals is NOT modified by this function.
+// removedPower - the total voting power that will be removed after the updates are verified and applied.
+//
 // Returns:
-// updatedTotalVotingPower - the new total voting power if these updates would be applied
-// numNewValidators - number of new validators
+// tvpAfterUpdatesBeforeRemovals -  the new total voting power if these updates would be applied without the removals.
+//   Note that this will be < 2 * MaxTotalVotingPower in case high power validators are removed and
+//   validators are added/ updated with high power values.
+//
+// numNewValidators -  number of new validators
 // err - non-nil if the maximum allowed total voting power would be exceeded
 //
-// 'updates' should be a list of proper validator changes, i.e. they have been verified
-// by processChanges for duplicates and invalid values.
-// No changes are made to the validator set 'vals'.
 func verifyUpdates(
 	updates []*Validator,
 	vals *ValidatorSet,
-) (updatedTotalVotingPower int64, numNewValidators int, err error) {
+	removedPower int64,
+) (tvpAfterUpdatesBeforeRemovals int64, numNewValidators int, err error) {
 
-	updatedTotalVotingPower = vals.TotalVotingPower()
+	// powerChanges is a temporary slice that eventually stores the changes, starting with the
+	// decreases in validator powers.
+	powerChanges := make([]powerChange, 0, len(updates))
```

Comment:
can't we just sort `updates` by `VotingPower` and do "compute the total voting power after updates starting with the power decreases" over them instead of introducing two new slices? This would radically simplify the code

Reply:
I had a version earlier doing that, needs the new sorting function and has higher computational complexity. If you feel strongly about it I can change it back, let me know.

Reply:
Is there a separate commit with those changes I can see?

Reply:
I just did a commit to use sort to see how it looks.

Reply:
Maybe I wasn't clear enough. I was talking about `updates`, not `powerChanges`. Why not sort `updates` and avoid having `powerChanges` altogether?

Reply:
Ah, sorry I missed that. Because `updates` need to be sorted by address for `applyUpdates()` that does the merge. So here we can sort updates by change but then sort them again by address.

Reply:
```
func verifyUpdates(
	updates []*Validator,
	vals *ValidatorSet,
	removedPower int64,
) (tvpAfterUpdatesBeforeRemovals int64, err error) {
	delta := func(update *Validator, vals *ValidatorSet) int64 {
		_, val := vals.GetByAddress(update.Address)
		if val != nil {
			return update.VotingPower - val.VotingPower
		}
		return update.VotingPower
	}

	updatesCopy := validatorListCopy(updates)
	sort.Slice(updatesCopy, func(i, j int) bool {
		return delta(updatesCopy[i], vals) < delta(updatesCopy[j], vals)
	})

	tvpAfterRemovals := vals.TotalVotingPower() - removedPower
	for _, upd := range updatesCopy {
		tvpAfterRemovals += delta(upd, vals)
		if tvpAfterRemovals > MaxTotalVotingPower {
			err = fmt.Errorf(
				"failed to add/update validator %v, total voting power would exceed the max allowed %v",
				upd.Address, MaxTotalVotingPower)
			return 0, err
		}
	}
	return tvpAfterRemovals + removedPower, nil
}
```

Reply:
The above function will benefit even more if we sort updates by delta from the beginning